### PR TITLE
Auth

### PR DESCRIPTION
**Pull request by worlvlhole**

### CODE REVIEW

File: internal/requests/middleware/verify_token.go [hunk 1]
```diff
@@ -0,0 +1,82 @@
...
+		return false
+	}
+
+	permsI, ok := rule.Configuration["permissions"]
+	if !ok {
+		logger.Error(policy.ErrNoPermission)
+		return false
+	}
+
+	perms, ok := permsI.([]interface{})
```
Comment (worlvlhole) [inline internal/requests/middleware/verify_token.go]: dup code

File: internal/token/policy/permission.go [hunk 2]
```diff
@@ -0,0 +1,106 @@
...
+	if r.Configuration == nil || len(r.Configuration) == 0 {
+		return ErrFailedToValidate
+	}
+
+	v, ok := r.Configuration["permissions"]
+	if !ok {
+		return ErrFailedToValidate
+	}
+
+	p, ok := v.([]interface{})
```
Comment (worlvlhole) [inline internal/token/policy/permission.go]: dup code

File: cmd/maladapt/main.go [hunk 3]
```diff
@@ -195,9 +207,18 @@ func main() {
 	r.Use(requests.NewStructuredLogger(requestLogger))
 
 	root := path.Join("/", config.VirtualPath)
+
 	//All apis should belong to this router
 	r.Route(root, func(r chi.Router) {
-		r.Mount("/file", fileAPI)
+		r.Group(func(r chi.Router) {
```
Comment (rosstimothy) [inline cmd/maladapt/main.go]: do we need two groups here?

File: internal/repository/token.go [hunk 4]
```diff
@@ -0,0 +1,188 @@
...
+			config,
+		)
+
+	default:
+		logger.Fatal("Database type not set")
+		return nil
+	}
+}
+
+//NewTokenMongoRepository creates a FileRepository backed by mongo
```
Comment (rosstimothy) [inline internal/repository/token.go]: typo `FileRepository`

File: internal/requests/middleware/verify_token.go [hunk 5]
```diff
@@ -0,0 +1,75 @@
...
+				next.ServeHTTP(w, r)
+				return
+			}
+
+			apiToken, err := repository.FindByToken(tokenHash[:])
+			if err != nil || apiToken == nil {
+				requests.WriteError(w, http.StatusUnauthorized, errors.New("invalid token").Error())
+				return
+			}
+			if !verifyTokenPermission(apiToken, r.URL.Path) {
```
Comment (rosstimothy) [inline internal/requests/middleware/verify_token.go]: Should this be its own middleware? Auth Middleware and Authz MIddleware?

File: internal/token/http_server.go [hunk 6]
```diff
@@ -0,0 +1,121 @@
...
+	defer func() {
+		span.End()
+		stats.Record(ctx, CreateCount.M(1))
+	}()
+
+	// Read body
+	b, err := ioutil.ReadAll(r.Body)
+	defer r.Body.Close()
+	if err != nil {
+		logger.Error(err)
```
Comment (rosstimothy) [inline internal/token/http_server.go]: stats

File: internal/token/http_server.go [hunk 7]
```diff
@@ -0,0 +1,121 @@
...
+		requests.WriteError(w, http.StatusInternalServerError, "")
+		return
+	}
+
+	var p policy.Policy
+	err = json.Unmarshal(b, &p)
+	if err != nil {
+		logger.Error(err)
```
Comment (rosstimothy) [inline internal/token/http_server.go]: stats

File: internal/token/service.go [hunk 8]
```diff
@@ -0,0 +1,100 @@
...
+	tokenRepo repository.TokenRepository
+}
+
+//NewService creates a token service with the given params
+func NewService(tokenRepo repository.TokenRepository) Service {
+	return &service{
+		tokenRepo: tokenRepo,
+	}
+}
+
```
Comment (rosstimothy) [inline internal/token/service.go]: //ProcessCreateRequest

File: internal/token/service.go [hunk 9]
```diff
@@ -0,0 +1,100 @@
...
+	err = s.storeToken(hashedToken[:], policies)
+	if err != nil {
+		logger.Error(err)
+		return "", err
+	}
+
+	return hex.EncodeToString(t), nil
+}
+
```
Comment (rosstimothy) [inline internal/token/service.go]: //ProcessDeleteRequest

File: internal/requests/middleware/verify_token.go (hunk 5 above)
Comment (worlvlhole) [reply internal/requests/middleware/verify_token.go]: yea.

**Pull request merged**